eyeball-im-util/src/vector/sort.rs

```diff
@@ -0,0 +1,613 @@
+    /// // | `ob`  | d b e f g a c |
+    /// // | `sub` | a b c d e f g |
+    /// //           ^   ^     ^^^
+    /// //           |   |     |
+    /// //           |   |     with `VectorDiff::Append { .. }`
+    /// //           |   with `VectorDiff::Insert { index: 2, .. }`
+    /// //           with `VectorDiff::PushFront { .. }`
+    ///
+    /// // Technically, `Sort` emits `VectorDiff`s that mimic a sorted `Vector`.
+    ///
+    /// drop(ob);
+    /// assert_closed!(sub);
+    /// # }
+    /// ```
+    ///
+    /// [`ObservableVector`]: eyeball_im::ObservableVector
+    #[project = SortProj]
+    pub struct Sort<S, F>
+    where
+        S: Stream,
+        S::Item: VectorDiffContainer,
+    {
+        // The main stream to poll items from.
+        #[pin]
+        inner_stream: S,
+
+        // The comparison function to sort items.
+        compare: F,
+
+        // This is the **sorted** buffered vector.
+        buffered_vector: Vector<(UnsortedIndex, VectorDiffContainerStreamElement<S>)>,
+
+        // This adapter can produce many items per item of the underlying stream.
+        //
+        // Thus, if the item type is just `VectorDiff<_>` (non-bached, can't
+        // just add diffs to a `poll_next` result), we need a buffer to store the
+        // possible extra items in.
+        ready_values: VectorDiffContainerStreamBuffer<S>,
+    }
+}
+
+impl<S, F> Sort<S, F>
+where
+    S: Stream,
+    S::Item: VectorDiffContainer,
+    F: Fn(&VectorDiffContainerStreamElement<S>, &VectorDiffContainerStreamElement<S>) -> Ordering
+        + Clone,
+{
+    /// Create a new `Filter` with the given (unfiltered) initial values, stream
+    /// of `VectorDiff` updates for those values, and filter.
+    pub fn new(
+        initial_values: Vector<VectorDiffContainerStreamElement<S>>,
+        inner_stream: S,
+        compare: F,
+    ) -> (Vector<VectorDiffContainerStreamElement<S>>, Self) {
+        let mut initial_values = initial_values.into_iter().enumerate().collect::<Vector<_>>();
+        initial_values.sort_by(|(_, left), (_, right)| compare(left, right));
+
+        (
+            initial_values.clone().into_iter().map(|(_, value)| value).collect(),
+            Self {
+                inner_stream,
+                compare,
+                buffered_vector: initial_values,
+                ready_values: Default::default(),
+            },
+        )
+    }
+
+    /// For testing purposes only.
+    #[doc(hidden)]
+    pub fn buffered_vector(&self) -> &Vector<(UnsortedIndex, VectorDiffContainerStreamElement<S>)> {
+        &self.buffered_vector
+    }
+}
+
+impl<S, F> Stream for Sort<S, F>
+where
+    S: Stream,
+    S::Item: VectorDiffContainer,
+    F: Fn(&VectorDiffContainerStreamElement<S>, &VectorDiffContainerStreamElement<S>) -> Ordering
+        + Clone,
+{
+    type Item = S::Item;
+
+    fn poll_next(self: Pin<&mut Self>, cx: &mut task::Context<'_>) -> Poll<Option<Self::Item>> {
+        self.project().poll_next(cx)
```

Comment:
This seems to be the only place that calls `SortProj::poll_next`. I think if no extra uses are intended to be introduced, that method should be inlined into this one, then the `#[project]` attribute on `Sort` can be removed too because the projected type doesn't need to be nameable.